corona-warn-app/cwa-app-android
# Alternative environment setup

## Description
* @ritsam  and me added a new environment setup where all necessary variables for each environment are defined in json map.
* The json map is loaded into a BuildConfig field and then parsed once at runtime.
* The environment variables (i.e. server urls and verification keys) are provided by `EnvironmentSetup` that has property to change which URLs are returned `currentEnvironment = PROD/DEV etc.`
* The test menu has a toggle that allows you to switch from the `DEFAULT` to the `ALTERNATIVE` (defined in the `build.gradle` file.
    * For the `device` buildFlavor this is always `PROD`/`PROD`, for the `deviceForTesters` flavor it is `DEV`/`WRU-XA`.
* If the local file `test_environments.json` does not exist, the checked in `prod_environments.json` is used.
* Switching the environment is only possible in `deviceForTesters` build as the toggle isn't otherwise available, the `EnvironmentSetup.currentEnvironment` does accept any changes in the `device` flavor.
 
## How to test
* Build the app in all different modes, and check that different environments are correctly used
* Build in `device` and in `deviceForTesters`, toggle the switch in test menu.
* Check logcat or via debugger that the correct URLs are used.

## Screenshot
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/54317407/94461163-aab8a980-01b1-11eb-9ed7-989e3ea1992e.gif)
